Subject: [PATCH] gitweb: incremental blame config

This adds support for a blame_incremental config option that rewrites
blame links to blame_incremental links when JavaScript is enabled.

It is disabled by default.  When enabled it only affects blame links
unlike javascript-actions.

Cc: Fredrik Kuivinen <frekui@gmail.com>
Signed-off-by: Petr Baudis <pasky@ucw.cz>
Signed-off-by: Kyle J. McKay <mackyle@gmail.com>
---
 gitweb/gitweb.perl | 47 +++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 41 insertions(+), 6 deletions(-)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index 74d45769..f5ffc7fb 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -320,6 +320,22 @@ our %feature = (
 		'override' => 0,
 		'default' => [0]},
 
+	# Enable the 'incremental blame' blob view, which uses javascript to
+	# incrementally show the revisions of lines as they are discovered
+	# in the history. It is better for large histories, files and slow
+	# servers, but requires javascript in the client and can slow down the
+	# browser on large files.
+
+	# To enable system wide have in $GITWEB_CONFIG
+	# $feature{'blame_incremental'}{'default'} = [1];
+	# To have project specific config enable override in $GITWEB_CONFIG
+	# $feature{'blame_incremental'}{'override'} = 1;
+	# and in project config gitweb.blame_incremental = 0|1;
+	'blame_incremental' => {
+		'sub' => sub { feature_bool('blame_incremental', @_) },
+		'override' => 0,
+		'default' => [0]},
+
 	# Enable the 'snapshot' link, providing a compressed archive of any
 	# tree. This can potentially generate high traffic if you have large
 	# project.
@@ -4154,6 +4170,16 @@ sub git_header_html {
 <meta name="generator" content="gitweb/$version git/$git_version$mod_perl_version"/>
 <meta name="robots" content="index, nofollow"/>
 <title>$title</title>
+<script type="text/javascript">/* <![CDATA[ */
+function fixBlameLinks() {
+	var allLinks = document.getElementsByTagName("a");
+	for (var i = 0; i < allLinks.length; i++) {
+		var link = allLinks.item(i);
+		if (link.className == 'blamelink')
+			link.href = link.href.replace("/blame/", "/blame_incremental/");
+	}
+}
+/* ]]> */</script>
 EOF
 	# the stylesheet, favicon etc urls won't work correctly with path_info
 	# unless we set the appropriate base URL
@@ -4255,6 +4281,9 @@ sub git_footer_html {
 
 		print qq!<script type="text/javascript">\n!.
 		      qq!window.onload = function () {\n!;
+		if (gitweb_check_feature('blame_incremental')) {
+			print qq!	fixBlameLinks();\n!;
+		}
 		if (gitweb_check_feature('javascript-actions')) {
 			print qq!	fixLinks();\n!;
 		}
@@ -4720,7 +4749,8 @@ sub git_print_tree_entry {
 		if ($have_blame) {
 			print " | " .
 			      $cgi->a({-href => href(action=>"blame", hash=>$t->{'hash'},
-			                             file_name=>"$basedir$t->{'name'}", %base_key)},
+						     file_name=>"$basedir$t->{'name'}", %base_key),
+				      -class => "blamelink"},
 			              "blame");
 		}
 		if (defined $hash_base) {
@@ -5006,7 +5036,8 @@ sub git_difftree_body {
 			              "blob") . " | ";
 			if ($have_blame) {
 				print $cgi->a({-href => href(action=>"blame", hash_base=>$parent,
-				                             file_name=>$diff->{'file'})},
+							     file_name=>$diff->{'file'}),
+					      -class => "blamelink"},
 				              "blame") . " | ";
 			}
 			print $cgi->a({-href => href(action=>"history", hash_base=>$parent,
@@ -5057,7 +5088,8 @@ sub git_difftree_body {
 			               "blob") . " | ";
 			if ($have_blame) {
 				print $cgi->a({-href => href(action=>"blame", hash_base=>$hash,
-				                             file_name=>$diff->{'file'})},
+							     file_name=>$diff->{'file'}),
+					      -class => "blamelink"},
 				              "blame") . " | ";
 			}
 			print $cgi->a({-href => href(action=>"history", hash_base=>$hash,
@@ -5103,7 +5135,8 @@ sub git_difftree_body {
 			              "blob") . " | ";
 			if ($have_blame) {
 				print $cgi->a({-href => href(action=>"blame", hash_base=>$hash,
-				                             file_name=>$diff->{'to_file'})},
+							     file_name=>$diff->{'to_file'}),
+					      -class => "blamelink"},
 				              "blame") . " | ";
 			}
 			print $cgi->a({-href => href(action=>"history", hash_base=>$hash,
@@ -7078,7 +7111,8 @@ sub git_blob {
 		if (defined $file_name) {
 			if ($have_blame) {
 				$formats_nav .=
-					$cgi->a({-href => href(action=>"blame", -replay=>1)},
+					$cgi->a({-href => href(action=>"blame", -replay=>1),
+						-class => "blamelink"},
 					        "blame") .
 					" | ";
 			}
@@ -8328,7 +8362,8 @@ XML
 			              -title => "diff"}, 'D');
 			if ($have_blame) {
 				print $cgi->a({-href => href(-full=>1, action=>"blame",
-				                             file_name=>$file, hash_base=>$commit),
+							     file_name=>$file, hash_base=>$commit),
+					      -class => "blamelink",
 				              -title => "blame"}, 'B');
 			}
 			# if this is not a feed of a file history
---
